### ISSUE TITLE
How to install (build, compile) zeromq for NW.js?

### ISSUE THREAD

**Issue by trusktr** [zeromq/zeromq.js#405]
It basically boils down to this:

```sh
set -e

# I'm using the `nw` package from NPM, and `nw-gyp` to rebuild native addons.
# Make sure `nw` is pinned if you want the sdk version (i.e. with no caret in
# package.json, otherwise it will install the non-sdk version because the
# version with a -sdk suffix treated as an in-range pre-release and thus means
# NPM will prefer to install the non-prerelease):
npm install nw@sdk --save-exact
# or for non-sdk, npm install nw
npm install nw-gyp --save-dev

# We need python 2. If our system's `python` is python3, we need to temporarily
# link it to python2 (install python2 if needed) to use nw-gyp.
# See https://github.com/nodejs/node/issues/2735 (you will get that error without this step).
# The following is on Arch/Manjaro Linux, but most likely your distro has a similar setup.
sudo ln -sf python2 /usr/bin/python

# configure and build zeromq with nw-gyp. Be sure to specify the correct NW.js
# version, and system architecture.
# See https://nwjs.readthedocs.io/en/latest/For%20Users/Advanced/Use%20Native%20Node%20Modules/#nw-gyp
pushd node_modules/zeromq
../.bin/nw-gyp configure --target=0.46.2 --python `which python2` --arch=x64
../.bin/nw-gyp build --target=0.46.2 --python `which python2` --arch=x64
popd

# restore `python` back to normal
sudo ln -sf python3 /usr/bin/python # on Arch/Manjaro Linux
```

**Comment by rolftimmermans** [comment]
I'm not familiar with NW.js, is there a particular reason that the native binaries cannot be used?

I don't know why nw-gyp would need Python 2, but the bundled gyp of Nodejs supports Python 3 nowadays.

**Comment by NbrTDB** [comment]
@rolftimmermans there are ABI differences that mean that native binaries don't necessarily work without rebuilding with nw-gyp: https://docs.nwjs.io/en/latest/For%20Users/Advanced/Use%20Native%20Node%20Modules/

nw-gyp wouldn't need Python 2, except that it doesn't appear to have been updated in a long while. There is a PR to bring it in line with node-gyp 6 if I recall.

**Comment by haviduck** [comment]
ive battled with this for 2 days now and im taking a ragequit break.  wanted to share my findings tho.

the first error i got when configuring was a sliced python error that really didnt make alot of sense due to the complete lack of smart logs. however, that can be fixed by editing the headers nw gyp downloads each time you pass a new target.  the gypi files in that directory needs a quick edit:
https://github.com/nwjs/node/issues/46#issue-764382891

ok so config finally went through but now at build im faced with a custom cxx compiler that probably was pointed at vs2017. solution there was triggering all the vars.bat i could find in 2015 installdirectory and mashing waaay to many paths to my env.  after that shameful move i fired up a new prompt (i use developer prompt for vs 2015, 17 etc) to ensure new paths was  caught. bam! cxx compiler approved.  after that it started throwing a link error. building via 2015 thats all it says. cant open file msvcrpt.lib. did some searches and saw that file was all around so i didnt get it and started really hating every single part of everything. thing is i dont even really need zmq in nwjs. theres so many other sockets i can go for. currently using sse for opencv streams, and that dinosaur of a protocol handles it well. BUT I WANT ZMQ IN MY APP BECAUSE IT ISNT WORKING.

i reached the rubberducking stage then finally i started tinkering with the what ifs. chose msvs_version 2017 and heres where i currently am. 2017 outputs that some of this stuff is x86 while im trying to compile for x64. so opened the sln up in vs and tried linking the x64 version of said file but nothing helped. im at best a novice vs guy and since i cant force a square to fit in a hole, ive taken up drinking and going back to  doing what i should been doing these last days.

electron on the other hand was no fuzz.

tldr;  the first two issues are fairly easy to beat albeit hacky. x86 on x64 compiling seems to be where its at. can someone educated come help and can i have a hug.
